[Ready for Review] Add Snapshot Workflow for Develop

hazel-sudz opened the pull request:
I looked at some tools and ended up deciding that just re-using the existing publish workflow makes the most sense without having to change the existing CI workflow for tests or causing a lot of diffs in general.

hazel-sudz committed:
add snapshot workflow

hazel-sudz commented on build.sbt:
```diff
@@ -118,7 +118,6 @@ val sharedSettings = Seq(
     setNextVersion,
     commitNextVersion,
     ReleaseStep(action = Command.process("sonatypeReleaseAll", _)),
-    pushChanges
   ),
```
todo: I will make this depend conditionally on if we are pushing a snapshot or not

hazel-sudz committed:
typo

hazel-sudz committed:
Setting version to 0.18.0-SNAPSHOT

hazel-sudz committed:
tmp

hazel-sudz committed:
update snapshot workflow

hazel-sudz committed:
remote version test file

hazel-sudz commented on build.sbt:
```diff
@@ -106,19 +106,18 @@ val sharedSettings = Seq(
+        publishArtifacts,
+        ReleaseStep(action = Command.process("sonatypeReleaseAll", _)),
+      )
+    // todo : implement workflow for final tagged releases that are not snapshots
+    else Seq[ReleaseStep](
+
+    )
```
this will be in follow-up PR once we have snapshots working well off develop

hazel-sudz replied:
I don't think we need to run tests as we already run tests on each PR so that should be fine

hazel-sudz committed:
update version.sbt to how it will work now

hazel-sudz commented on version.sbt:
```diff
@@ -1 +1 @@
-ThisBuild / version := "0.18.0-SNAPSHOT"
+ThisBuild / version := "0.18.0"
```
on snapshot publish, this will get prepended with SNAPSHOT by the CI. This is also how we can differentiate snapshot vs non-snapshot release on build.sbt file.

johnynek replied:
I wonder if we can use:

https://github.com/sbt/sbt-dynver

hazel-sudz replied:
sounds fine to me will implement I like the "fetch from recent tag" idea.

hazel-sudz commented:
I need the following loaded into the repo secrets to test this any further: 

| Github Repo Secret  | Used by Plugin |
| ------------- | ------------- |
| SONATYPE_PASSWORD  | https://github.com/xerial/sbt-sonatype |
| SONATYPE_USERNAME  | https://github.com/xerial/sbt-sonatype |
| PGP_PASSPHRASE  |https://github.com/sbt/sbt-pgp |
| PGP_SECRET  | https://github.com/sbt/sbt-pgp |

hazel-sudz committed:
do not use inquireVersions to be automated

hazel-sudz committed:
load env secrets

hazel-sudz committed:
use dynver & add support for github releases

hazel-sudz commented on build.sbt:
```diff
@@ -102,25 +102,26 @@ val sharedSettings = Seq(
   releasePublishArtifactsAction := PgpKeys.publishSigned.value,
-  releaseVersionBump := sbtrelease.Version.Bump.Minor, // need to tweak based on mima results
+  ThisBuild / dynverSonatypeSnapshots := true, // prepend "-SNAPSHOT" to version tag when releasing a snapshot style build
+  ThisBuild / dynverSeparator := "-", // use a URI friendly separator for snapshots instead of '+' char
   publishMavenStyle := true,
   Test / publishArtifact := false,
   pomIncludeRepository := { x => false },
-  releaseProcess := Seq[ReleaseStep](
-    checkSnapshotDependencies,
-    inquireVersions,
-    runClean,
-    runTest,
-    setReleaseVersion,
-    commitReleaseVersion,
-    tagRelease,
-    publishArtifacts,
-    setNextVersion,
-    commitNextVersion,
-    ReleaseStep(action = Command.process("sonatypeReleaseAll", _)),
-    pushChanges
-  ),
```
any kind of versioning through `version.sbt` file is now no longer needed thanks to https://github.com/sbt/sbt-dynver

final versions can now be trigged trough the Github Releases UI in which you can select a branch and a tag.

hazel-sudz commented on version.sbt:
```diff
@@ -1 +0,0 @@
-ThisBuild / version := "0.18.0-SNAPSHOT"
```
`develop` currently resolves to `0.16.1-RC2` based on tags in the git history. Would need to push a new tag for the snapshot versions to start making sense

johnynek replied:
yeah, I can take that. https://search.maven.org/artifact/com.twitter/scalding-core_2.12 looks like 0.18.0-RC1 was the last published.

hazel-sudz committed:
attempt to fix tests that relied on version.sbt

johnynek commented:
https://github.com/sbt/sbt-buildinfo

if we add that, we can remove any code that depends on version in sbt and instead get it from a generated object.

johnynek commented:
actually using this we can also add the gitsha:

https://brainchunk.blogspot.com/2016/11/embed-git-commit-hash-scala-project.html;

hazel-sudz committed:
override sbd-dynver

hazel-sudz commented:
> actually using this we can also add the gitsha:
> 
> https://brainchunk.blogspot.com/2016/11/embed-git-commit-hash-scala-project.html;

gitsha can be handled by `sbt-dynver`. It is written in the version by default if the HEAD commit does not have a tag (ie. for SNAPSHOTS)

hazel-sudz commented:
> https://github.com/sbt/sbt-buildinfo
> 
> if we add that, we can remove any code that depends on version in sbt and instead get it from a generated object.

++ I think there's one place doing something with scalding version (but hasn't been updated in a while so probably not critical?). I'm currently having some problems with scald.rb not finding the proper versioned artifacts working on resolving that.

hazel-sudz committed:
fix scalding version in scald.rb

hazel-sudz committed:
cherry-pick CI fixes for dynamic versioning to work

no tag so should drop to snapshot against 2.0 with commit sha

fix versioning

remove all debug diffs

no format on execuation tutorial

do not format logs

debug with reverse ssh

add more logging info

cd before resolving scalding version

remove debug assert

try like this

fix syntax

fix syntax

debug - load tags

more debug

debug

hazel-sudz commented:
was fixing some test issues caused by the dynamic version. Will add changes in a bit. Here is what the versioning behaviour will look like afterwards:


```scala
// head commit has been tagged with v2.0
dsudzilouski@dsudzilouski scalding % ./sbt -Dsbt.log.noformat=true -Dsbt.supershell=false "print scalding-core / version" -error
2.0

// if head commit is tagged should always assume a final release even if git history is dirty
dsudzilouski@dsudzilouski scalding % touch TEST_FILE_DIRTY
dsudzilouski@dsudzilouski scalding % git add .
dsudzilouski@dsudzilouski scalding % git status
On branch debug_CI
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   TEST_FILE_DIRTY

dsudzilouski@dsudzilouski scalding % ./sbt -Dsbt.log.noformat=true -Dsbt.supershell=false "print scalding-core / version" -error
2.0

// head commit is untagged, should assume SNAPSHOT release
dsudzilouski@dsudzilouski scalding % git commit -m "no tag so should drop to snapshot against 2.0 with commit sha"
[debug_CI 7ed1fdbb] no tag so should drop to snapshot against 2.0 with commit sha
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 TEST_FILE_DIRTY

dsudzilouski@dsudzilouski scalding % ./sbt -Dsbt.log.noformat=true -Dsbt.supershell=false "print scalding-core / version" -error
2.0-1-7ed1fdbb-SNAPSHOT
```

hazel-sudz committed:
add diff for reference in PR

hazel-sudz commented on scalding-core/src/main/scala/com/twitter/package.scala:
```diff
@@ -34,7 +34,7 @@ package object scalding {
    * Make sure this is in sync with version.sbt
    */
-  val scaldingVersion: String = "0.17.2"
+  val scaldingVersion: String = "0.18.0"
```
I believe this is the only place that we refer to scalding version within src. ~~As mentioned above, this can be addressed with something like https://github.com/sbt/sbt-buildinfo. But considering that it is already out of date and clearly not critical, I think it would be fair for a followup~~ (done) CC @johnynek

hazel-sudz replied:
I have moved this change to https://github.com/twitter/scalding/pull/1976 as it is independent of what is being proposed here

hazel-sudz commented:
@nownikhil any chance we could have help with creds being imported to the repo secrets section?

nownikhil commented:
Apparently both me and Navin don't have the right permissions. There is twitter internal process to get access to publish to external repo. @navinvishy is working on that.

johnynek commented:
can you merge in develop?

hazel-sudz committed:
Merge branch 'develop' into publish_ci

daniel-sudz commented:
we can merge it's just this will fail until we have the publish secrets available. It's probably not a big deal (and also might be helpful to sort other bugs if there any).

hazel-sudz commented:
@nownikhil are there any updates on publish creds? We would really love to have some RC artifacts to do integration with real production jobs.

nownikhil commented:
I got the impression that Oscar (@johnynek) is gonna setup up the secrets since he is the admin. He commented on https://github.com/twitter/scalding/issues/1973

I will follow up on the internal process. The internal ticket is still unassigned, let me ping some folks here and see if they know any better.

johnynek commented:
I can take this. Sorry I just haven't gotten to it.
